src/lib/tree/tree.md

```diff
@@ -0,0 +1,90 @@
+with `mat-` instead of `cdk-`.
+
+There are two types of trees: Flat tree and Nested Tree. The DOM structures are different for these
+two types of trees.
+
+#### Flat tree
+Flat tree has a list of flattened tree nodes. The hierarchy are flattened and no tree node is placed
+inside another tree node. The hierarchy information of the tree is kept in `level`.
+We will need a `TreeFlattener` to flat the tree whenever the tree node changes or
+expanded/collapsed.
+
+```
+<mat-tree>
+    <mat-tree-node> parent node </mat-tree-node>
+    <mat-tree-node> -- child node1 </mat-tree-node>
+    <mat-tree-node> -- child node2 </mat-tree-node>
+</mat-tree>
+```
+A flat tree is easier to implement virtual scrolling in the future or render only part of the tree.
+
+<!--TODO(tinayuangao): Add a flat tree example here -->
+
+#### Nested tree
+In Nested tree, children nodes are placed inside their parent node in DOM. The parent node has an
+outlet to keep all the children nodes.
+
+```
+<mat-tree>
+   <mat-nested-tree-node>
+       parent node
+       <mat-nested-tree-node> -- child node1 </mat-tree-node>
+       <mat-nested-tree-node> -- child node2 </mat-tree-node>
+   </mat-nested-tree-node>
+</mat-tree>
+```
+
+A nested tree is easier to decorate to show hierarchy information.
+
+<!--TODO(tinayuangao): Add a nested tree example here -->
+
+### Features
+
+The `<mat-tree>` itself only deals with the rendering of a tree structure.
+Additional features can be built on top of the tree by adding behavior inside node templates
+(e.g., padding and trigger). Interactions that affect the
+rendered data (such as expand/collapse) should be propagated through the table's data source.
+
+### TreeControl
+
+The `TreeControl` controls the expand/collapse state of tree nodes. Users can expand/collapse a tree
+node recursively through tree control. For nested tree node, `getChildren` function need to pass to
+the `NestedTreeControl` to make it work recursively. For flattened tree node, `getLevel` and
+`isExpandable` functions need to pass to the `FlatTreeControl` to make it work recursively.
+ 
+### Trigger
+
+A `matTreeNodeTrigger` can be added in the tree node template to expand/collapse the tree node. The
+trigger triggers the expand/collapse functions in `TreeControl` and is able to expand/collapse a 
+tree node recursively by setting `[matTreeNodeTriggerRecursive]` to `true`. 
+
+The trigger can be placed anywhere in the tree node, and is only triggered by `click` action. 
+
+
+### Padding (Flat tree only)
+
+The `matTreeNodePadding` can be placed in a flat tree's node template to display the `level` 
+information of a flat tree node.
+
+Nested tree does not need this padding since padding can be easily added to the hierarchy
+structure in DOM.
+
+### Simple Tree
+
+In the near future, we will provide a simplified version of the tree with an easy-to-use
+interface, material styling, and json input.
```

Comment:
I would leave out this section; I want to avoid talking about future stuff in the docs